File: 03 - source code/01 - imperative languages/C#/random_streams_for_perf_stats.cs
/*
random_streams_for_perf_stats.cs

2025-06-07

build on Ubuntu 24 LTS: $ dotnet new console -n random_streams_for_perf_stats --use-program-main
                        $ cd ./random_streams_for_perf_stats
                        development candidate:
                        $ dotnet build

run on Ubuntu 24 LTS:   $ ./bin/Debug/net8.0/random_streams_for_perf_stats

                        release candidate:
                    (A) $ dotnet publish -c Release -r linux-x64 --no-self-contained   <<<<<<<<<<<<<<

                    (B) $ dotnet publish -c Release -r linux-x64 --self-contained true -p:PublishSingleFile=true -p:PublishReadyToRun=true
                        self-contained == not requiring .NET Core runtime installed on host machine
                        => this makes a slower program: ~+20%


run on Ubuntu 24 LTS:   $ ./bin/Release/net8.0/linux-x64/random_streams_for_perf_stats
                        $ sudo perf stat -r 20 ./bin/Release/net8.0/linux-x64/random_streams_for_perf_stats


(Using AOT:             $ https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=linux-ubuntu%2Cnet8
                        --> add to: random_streams_for_perf_stats.csproj
                                    <PublishAot>true</PublishAot>

                    (A) $ dotnet publish -c Release -r linux-x64
                        $ ./bin/Release/net8.0/linux-x64/random_streams_for_perf_stats
                        $ sudo perf stat -r 20 ./bin/Release/net8.0/linux-x64/random_streams_for_perf_stats

                    (B) $ dotnet publish -c Release -r linux-x64 --self-contained true -p:PublishSingleFile=true -p:PublishReadyToRun=true
                        $ ./bin/Release/net8.0/linux-x64/random_streams_for_perf_stats
                        $ sudo perf stat -r 20 ./bin/Release/net8.0/linux-x64/random_streams_for_perf_stats
                        => this makes a slower program: ~+20%
)


(a) install the .NET runtime:
$ wget https://dot.net/v1/dotnet-install.sh -O dotnet-install.sh
$ ./dotnet-install.sh --version latest
https://learn.microsoft.com/en-us/dotnet/core/install/linux-scripted-manual#scripted-install

$ dotnet --list-runtimes
  Microsoft.AspNetCore.App 8.0.4 [/usr/lib/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 8.0.4 [/usr/lib/dotnet/shared/Microsoft.NETCore.App]

(b) install the .NET SDK:
$ sudo apt-get update && sudo apt-get install -y dotnet-sdk-8.0
$ dotnet --list-sdks
8.0.104 [/usr/lib/dotnet/sdk]


*/

namespace random_streams_for_perf_stats;

using System.Text;  // StringBuilder
using System.IO;    // File

class Program
{
  static void Main(string[] args)
  {
    const int END = 62501;  // 62501 for exactly 1M binary digits
    // const int END  = 15;  // for testing
    const int M1   = END * 16;
    const int K250 = END * 4;

    const int m = 65521;  // = 2^16 - 15
    const int a = 17364;
    const int c = 0;

    const string file_bits_x   = "random_bitstring.bin";
    const string file_bits_hex = "random_bitstring.byte";


    int[] x = new int[END];  // arrays cannot be resized dynamically

    Random rand = new Random();
    x[0] = rand.Next(1, m);
    // Console.WriteLine("x[0] = " + x[0]);  // for testing

    StringBuilder bits_x   = new StringBuilder(M1);
    StringBuilder bits_hex = new StringBuilder(K250);


    Console.WriteLine("\ngenerating a random bit stream...");
    for (int i = 1; i < END; i = i + 1) {
      x[i] = (a * x[i-1] + c) % m;
      // Console.WriteLine("\n" + x[i]);  // for testing

      string bits_x_str = String.Format("{0,16:B16}", x[i]);
      // Console.WriteLine(bits_x_str);  // for testing
      bits_x.Append(bits_x_str);

      string bits_hex_str = String.Format("{0,4:x4}", x[i]);
      // Console.WriteLine(bits_hex_str);  // for testing
      bits_hex.Append(bits_hex_str);
    }

    // write bit stream to disk:
    try {
      File.WriteAllText(file_bits_x, bits_x.ToString());
      Console.WriteLine("Bit stream has been written to disk under name:  " + file_bits_x);
    } catch (Exception ex) {
      Console.WriteLine($"could not write to file: {file_bits_x} ! -- {ex.Message}");
    }

    // write byte stream to disk:
    try {
      File.WriteAllText(file_bits_hex, bits_hex.ToString());
      Console.WriteLine("Byte stream has been written to disk under name: " + file_bits_hex);
    } catch (Exception ex) {
      Console.WriteLine($"could not write to file: {file_bits_hex} ! -- {ex.Message}");
    }

  }
}

// end of random_streams_for_perf_stats.c3.cs

File: 03 - source code/01 - imperative languages/Java/RandomStreamsForPerfStats.java
/*
RandomStreamsForPerfStats.java

2026-05-14/18

run on Ubuntu 24 LTS:   $ time java RandomStreamsForPerfStats.java

                        generating a random bit stream...
                        Bit stream has been written to disk under name:  random_bitstring.bin
                        Byte stream has been written to disk under name: random_bitstring.byte

                        real	0m0.357s <<<<<<<<<<<<<<<
                        user	0m1.041s
                        sys	0m0.060s
                        $

build on Ubuntu 24 LTS: $ javac RandomStreamsForPerfStats.java
                        $ jar cfev RandomStreamsForPerfStats.jar RandomStreamsForPerfStats RandomStreamsForPerfStats.class
                        # switch e is essential, see from here: https://docs.oracle.com/javase/tutorial/deployment/jar/appman.html

                        $ time java -jar RandomStreamsForPerfStats.jar

                        generating a random bit stream...
                        Bit stream has been written to disk under name:  random_bitstring.bin
                        Byte stream has been written to disk under name: random_bitstring.byte

                        real	0m0.052s <<<<<<<<<<<<<<<
                        user	0m0.103s
                        sys	0m0.023s
                        $


$ java --version
openjdk 25.0.2 2026-01-20
OpenJDK Runtime Environment (build 25.0.2+10-Ubuntu-124.04)
OpenJDK 64-Bit Server VM (build 25.0.2+10-Ubuntu-124.04, mixed mode, sharing)
$

mostly  transpiled from RandomStreamsForPerfStats.hx (Haxe) with Duck.ai (using GPT-5 mini) and Google AI

*/

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.concurrent.ThreadLocalRandom;

public class RandomStreamsForPerfStats {
    public static void main() {
        final int END = 62501;

        final int m = 65521;
        final int a = 17364;
        final int c = 0;

        final String fileBitsX   = "random_bitstring.bin";
        final String fileBitsHex = "random_bitstring.byte";

        int[] x = new int[END];

        // Std.random(m - 1) + 1  -> random in [1, m-1]
        x[0] = ThreadLocalRandom.current().nextInt(1, m);

        StringBuilder bitsX = new StringBuilder();
        StringBuilder bitsHex = new StringBuilder();

        System.out.println("\ngenerating a random bit stream...");
        for (int i = 1; i < END; i++) {
            x[i] = (int)(((long)a * x[i - 1] + c) % m);

            // Convert to binary string and pad with leading zeros up to 16 characters (Google AI)
            String bitsXStr = String.format("%16s", Integer.toBinaryString(x[i])).replace(' ', '0');
            bitsX.append(bitsXStr);

            String bitsHexStr = String.format("%4s", Integer.toHexString(x[i])).replace(' ', '0');
            bitsHex.append(bitsHexStr);
        }

        String bitsXStrTotal = bitsX.toString();
        String bitsHexStrTotal = bitsHex.toString();

        // write bit stream to disk:
        try {
            Files.writeString(Path.of(fileBitsX), bitsXStrTotal, StandardCharsets.UTF_8);
            System.out.println("Bit stream has been written to disk under name:  " + fileBitsX);
        } catch (IOException ex) {
            System.out.println("could not write to file: " + fileBitsX + " ! -- " + ex);
        }

        // write byte stream to disk:
        try {
            Files.writeString(Path.of(fileBitsHex), bitsHexStrTotal, StandardCharsets.UTF_8);
            System.out.println("Byte stream has been written to disk under name: " + fileBitsHex);
        } catch (IOException ex) {
            System.out.println("could not write to file: " + fileBitsHex + " ! -- " + ex);
        }
    }
}

// end of RandomStreamsForPerfStats.java
